apollo-federation/src/connectors/json_selection/methods/public/eq.rs

```diff
@@ -48,17 +72,53 @@ fn eq_method(
 #[allow(dead_code)] // method type-checking disabled until we add name resolution
 fn eq_shape(
     method_name: &WithRange<String>,
-    _method_args: Option<&MethodArgs>,
-    _input_shape: Shape,
-    _dollar_shape: Shape,
-    _named_var_shapes: &IndexMap<&str, Shape>,
+    method_args: Option<&MethodArgs>,
+    input_shape: Shape,
+    dollar_shape: Shape,
+    named_var_shapes: &IndexMap<&str, Shape>,
     source_id: &SourceId,
 ) -> Shape {
+    let arg_count = method_args.map(|args| args.args.len()).unwrap_or_default();
+    if arg_count > 1 {
+        return Shape::error(
+            format!(
+                "Method ->{} requires only one argument, but {arg_count} were provided",
+                method_name.as_ref(),
+            ),
+            vec![],
+        );
+    }
+
+    let Some(first_arg) = method_args.and_then(|args| args.args.first()) else {
+        return Shape::error(
+            format!("Method ->{} requires one argument", method_name.as_ref()),
+            method_name.shape_location(source_id),
+        );
+    };
+    let arg_shape = first_arg.compute_output_shape(
+        input_shape.clone(),
+        dollar_shape,
+        named_var_shapes,
+        source_id,
+    );
+
+    // Ensures that the arguments are of the same type... this includes covering cases like int/float and unknown/name
+    if !(input_shape.accepts(&arg_shape) || arg_shape.accepts(&input_shape)) {
+        return Shape::error_with_partial(
+            format!(
+                "Method ->{} requires the applied to value and argument to be the same type to be comparable.",
```

Comment:
```suggestion
                "Method ->{} can only compare values of the same type.",
```

"applied to" is very internal terminology